bigtable/client/internal/row_reader.h

```diff
@@ -0,0 +1,158 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ROW_READER_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_INTERNAL_ROW_READER_H_
+
+#include "bigtable/client/version.h"
+
+#include "bigtable/client/data_client.h"
+#include "bigtable/client/filters.h"
+#include "bigtable/client/internal/readrowsparser.h"
+#include "bigtable/client/row.h"
+#include "bigtable/client/row_set.h"
+#include "bigtable/client/rpc_backoff_policy.h"
+#include "bigtable/client/rpc_retry_policy.h"
+
+#include <google/bigtable/v2/bigtable.grpc.pb.h>
+
+#include <absl/memory/memory.h>
+#include <absl/types/optional.h>
+#include <grpc++/grpc++.h>
+
+#include <iterator>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Object returned by Table::ReadRows(), enumerates rows in the response.
+ *
+ * Read rows can be obtained by iterating on this object via begin()
+ * and end().  Once an iterator is incremented, all data in the row it
+ * initially pointed to is freed.
+ *
+ */
+class RowReader {
+  class RowReaderIterator;
+
+ public:
+  /// Signifies that there is no limit on the number of rows to read.
+  static int const NO_ROWS_LIMIT = 0;
+
+  RowReader(std::shared_ptr<DataClient> client, absl::string_view table_name,
+            RowSet row_set, int rows_limit, Filter filter,
+            std::unique_ptr<RPCRetryPolicy> retry_policy,
+            std::unique_ptr<RPCBackoffPolicy> backoff_policy);
+
+  using iterator = RowReaderIterator;
+
+  /**
+   * Input iterator over rows in the response.
+   *
+   * All iterators that do not point to end() are equivalent, that is,
+   * incrementing one iterator will increment all by advancing the
+   * stream.
+   */
+  iterator begin();
+
+  /// End iterator over the rows in the response.
+  iterator end();
+
+  /**
+   * Stop the read call and clean up the connection.
+   *
+   * Invalidates iterators.
+   */
+  void Cancel();
+
+ private:
+  /**
+   * Read and parse the next row in the response.
+   *
+   * Invalidates the previous row and feeds response chunks into the
+   * parser until another row is available.
+   *
+   * This call possibly blocks waiting for data.
+   */
+  void Advance();
+
+  /// Called by Advance(), does not handle retries.
+  grpc::Status AdvanceOrFail();
+
+  /**
+   * Move the index to the next chunk, reading data if needed.
+   *
+   * Returns false if no more chunks are available.
+   */
+  bool NextChunk();
+
+  /// Sends the ReadRows request to the stub.
+  void MakeRequest();
+
+  /// The input iterator returned by begin() and end()
+  class RowReaderIterator : public std::iterator<std::input_iterator_tag, Row> {
+   public:
+    RowReaderIterator(RowReader* owner, bool is_end)
+        : owner_(owner), is_end_(is_end) {}
+
+    RowReaderIterator& operator++();
```

Comment:
I think you are required to implement the postfix `++` operator too, though it might be a bad idea to use it.

Reply:
thanks, I followed your advice and refactored to hold the row in the iterator.

I kept to the absl::optional to be able to use emplace.

Updated the * and -> operators as well by basically snatching the ones from absl::optional.